src/main/java/graphql/annotations/dataFetchers/MethodDataFetcher.java

```diff
@@ -118,40 +132,67 @@ private Object[] invocationArgs(DataFetchingEnvironment environment, ProcessingE
         return result.toArray();
     }
 
+    @SuppressWarnings("ConstantConditions")
     private Object buildArg(Type p, GraphQLType graphQLType, Object arg) {
-        if (arg == null) {
+        Optional<Object> optionalArg = Optional.ofNullable(arg);
+        if (!optionalArg.isPresent()) {
             return null;
         }
         if (graphQLType instanceof graphql.schema.GraphQLNonNull) {
             graphQLType = ((graphql.schema.GraphQLNonNull) graphQLType).getWrappedType();
         }
+
         if (p instanceof Class<?> && graphQLType instanceof GraphQLInputObjectType) {
-            Constructor<?> constructors[] = ((Class) p).getConstructors();
+            Constructor<?>[] constructors = ((Class) p).getConstructors();
             Constructor<?> constructor = getBuildArgConstructor(constructors);
             Parameter[] parameters = constructor.getParameters();
-            if (parameters.length == 1 && parameters[0].getType().isAssignableFrom(arg.getClass())) {
-                return constructNewInstance(constructor, arg);
+
+            if (parameters.length == 1 && optionalArg.isPresent() && parameters[0].getType().isAssignableFrom(optionalArg.get().getClass())) {
+                if (parameters[0].getType().isAssignableFrom(Optional.class)) {
+                    return constructNewInstance(constructor, arg);
+                } else {
+                    return constructNewInstance(constructor, optionalArg.orElse(null));
+                }
```

Comment:
I don't see how this case can work. The constructor has one single parameter which is assignable from `arg.getClass()` (Map, for example) and at the same time Optional.class , and in that case , you don't pass the optional but the arg ? otherwise, you keep the previous behaviour , equivalent to `return constructNewInstance(constructor, arg);` ? In that case, rather keep the previous code